File: assent/policy.py
"""The deterministic policy engine — the trust decision, in code.

This is step 5 of the pipeline in ``docs/policy-engine.md``: a pure function mapping
``(risk_envelope, owner)`` — plus the catalog and rollback facts carried on the
``Change`` — to ``{auto | route-to-owner | escalate}``. It is deterministic on
purpose: the trust decision "must be code: auditable, testable, versioned."

The invariant (D6) is enforced structurally here, not by convention:

* Every gate-*opening* check reads a measured fact (catalog, environment, blast radius,
  reversibility, ownership graph). None reads model confidence.
* ``confidence`` and ``context_caution`` are consulted only in the *downgrade* pass,
  which can move a decision toward ``ESCALATE`` but never toward ``AUTO``.

So auto-execution is earned by a **low risk envelope**, never by **high confidence** —
the sharp break from "99% confidence => act."
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import List

from assent.catalog import ActionCatalog, DEFAULT_CATALOG
from assent.change import Change, Environment, Reversibility


class Decision(str, Enum):
    """The three terminal gates. Ordered from most to least autonomous so that a
    downgrade is always a move to a later member."""

    AUTO = "auto"                      # measured envelope is low enough to auto-execute
    ROUTE_TO_OWNER = "route_to_owner"  # gate, sent to the authoritative owner
    ESCALATE = "escalate"              # gate, broadened — no confident owner / fail-safe

    @property
    def rank(self) -> int:
        return {"auto": 0, "route_to_owner": 1, "escalate": 2}[self.value]


@dataclass(frozen=True)
class PolicyResult:
    """The engine's output: the decision plus the audit trail of why."""

    decision: Decision
    reasons: List[str] = field(default_factory=list)

    def with_decision(self, decision: Decision, reason: str) -> "PolicyResult":
        return PolicyResult(decision=decision, reasons=[*self.reasons, reason])


@dataclass(frozen=True)
class AutonomyPolicy:
    """The per-customer *dial*. Autonomy is earned, so the defaults are conservative:
    no auto-execution in prod, only reversible actions, tight blast radius, and a
    confidence floor below which we always escalate.

    None of these fields can *open* a gate the measured facts have closed — they only
    make the AUTO bar harder to clear.
    """

    allow_auto_writes: bool = True             # master switch for any write autonomy
    auto_environments: frozenset = frozenset({Environment.DEV, Environment.STAGING})
    max_auto_blast_radius: int = 5             # narrow only
    allow_auto_tier0: bool = False             # crown-jewel targets never auto
    min_owner_confidence: float = 0.75         # below => cannot silently route
    min_confidence_for_auto: float = 0.85      # confidence can only tighten, never open


class PolicyEngine:
    """Deterministic gating. Construct once per policy version; ``evaluate`` is pure."""

    def __init__(
        self,
        catalog: ActionCatalog = DEFAULT_CATALOG,
        autonomy: AutonomyPolicy = AutonomyPolicy(),
    ) -> None:
        self.catalog = catalog
        self.autonomy = autonomy

    def evaluate(self, change: Change) -> PolicyResult:
        env = change.risk_envelope
        action = change.action
        reasons: List[str] = []

        # --- Step 1: catalog is the safety boundary. Unknown => fail safe to human. ---
        action_class = self.catalog.get(action.type)
        if action_class is None:
            return PolicyResult(
                Decision.ESCALATE,
                [f"action '{action.type}' is not in the catalog; failing safe to human"],
            )

        # --- Step 2: reads are autonomous. The real trust boundary is read vs write. ---
        if not action.is_write:
            return PolicyResult(Decision.AUTO, ["read-only action; reads are autonomous"])

        # From here down we are gating a *write*. Decide whether the measured facts
        # permit AUTO. Any single failure closes the AUTO door; we then choose between
        # ROUTE_TO_OWNER and ESCALATE based on ownership confidence.

        auto_blocked: List[str] = []

        if not self.autonomy.allow_auto_writes:
            auto_blocked.append("write autonomy is disabled for this customer")

        # A write with no rollback plan can never auto-execute. "No undo => no autonomy."
        if not change.has_rollback:
            auto_blocked.append("no rollback plan; a write without undo never auto-executes")

        # Reversibility is measured from the catalog *and* the concrete envelope; take
        # the more conservative of the two.
        effective_reversibility = _most_conservative(
            action_class.reversibility, env.reversibility
        )
        if effective_reversibility is not Reversibility.REVERSIBLE:
            auto_blocked.append(
                f"reversibility is '{effective_reversibility.value}'; only reversible "
                f"writes auto-execute"
            )

        if env.environment not in self.autonomy.auto_environments:
            auto_blocked.append(
                f"environment '{env.environment.value}' is outside the auto set"
            )

        if env.hits_tier0 and not self.autonomy.allow_auto_tier0:
            auto_blocked.append("target is tier-0; tier-0 writes never auto-execute")

        if env.blast_radius > self.autonomy.max_auto_blast_radius:
            auto_blocked.append(
                f"blast_radius {env.blast_radius} exceeds auto ceiling "
                f"{self.autonomy.max_auto_blast_radius}"
            )

        if auto_blocked:
            # Measured facts closed the AUTO door. Route if we trust the owner, else escalate.
            return self._gate(change, reasons=auto_blocked)

        # --- Downgrade pass: model-supplied signals may only TIGHTEN, never open. ---
        # At this point the measured envelope permits AUTO. Confidence and context can
        # still push us to a human, but nothing here could have opened a closed gate.
        if env.confidence < self.autonomy.min_confidence_for_auto:
            return self._gate(
                change,
                reasons=[
                    f"confidence {env.confidence:.2f} below auto floor "
                    f"{self.autonomy.min_confidence_for_auto:.2f}; confidence only tightens"
                ],
            )

        if change.context_caution:
            return self._gate(
                change,
                reasons=["doc-grounded caution flag raised; context tightens, never opens"],
            )

        return PolicyResult(
            Decision.AUTO,
            [
                "reversible + narrow + in-scope environment with a rollback plan and "
                "sufficient confidence; low risk envelope earns auto-execution"
            ],
        )

    def _gate(self, change: Change, reasons: List[str]) -> PolicyResult:
        """AUTO is off the table; choose the human path.

        A confident owner gets the change routed straight to them. An unknown or
        low-confidence owner is never silently routed — it escalates (broadens),
        because a stale/guessed owner is a dangerous approver.
        """
        owner = change.owner
        if owner.known and owner.confidence >= self.autonomy.min_owner_confidence:
            return PolicyResult(
                Decision.ROUTE_TO_OWNER,
                [*reasons, f"routing to owner '{owner.id}' (confidence {owner.confidence:.2f})"],
            )
        return PolicyResult(
            Decision.ESCALATE,
            [
                *reasons,
                "owner unknown or below routing-confidence floor; escalating rather than "
                "guessing an approver",
            ],
        )


def _most_conservative(a: Reversibility, b: Reversibility) -> Reversibility:
    """Return the more conservative (less reversible) of two classifications."""
    order = {
        Reversibility.REVERSIBLE: 0,
        Reversibility.RECOVERABLE: 1,
        Reversibility.IRREVERSIBLE: 2,
    }
    return a if order[a] >= order[b] else b

File: src/policy.ts
/**
 * The policy engine — the moat.
 *
 * `policy()` is a pure, deterministic function of the risk envelope, the owner,
 * and a few measured facts about the change. It IS the trust decision, so it is
 * code: auditable, testable, versioned (docs/decisions.md — D2, D6).
 *
 * The single invariant it enforces (docs/policy-engine.md):
 *
 *   The LLM may produce any input that only makes the gate MORE conservative.
 *   Anything that could RELAX the gate must be deterministic.
 *
 * Concretely, auto-execution is earned by a LOW RISK ENVELOPE (reversible +
 * narrow + non-prod) — never by high confidence. Confidence and doc-grounded
 * context can only tighten.
 */

import {
  BlastRadius,
  Confidence,
  Decision,
  Effect,
  Environment,
  type PolicyResult,
  Reversibility,
} from "./types.ts";

/**
 * The autonomy dial. Autonomy is earned per customer, not a launch default
 * (docs/objectives.md, roadmap Phase 3). With `writesMayAutoExecute: false`
 * (the safe default), every write gates and only reads run autonomously.
 */
export interface AutonomyDial {
  /** Master switch for auto-executing writes. Default: false. */
  writesMayAutoExecute: boolean;
  /** Widest blast radius still eligible for auto-execution. */
  maxAutoBlastRadius: BlastRadius;
  /** Environments in which writes may auto-execute. Prod is never included. */
  autoEnvironments: ReadonlySet<Environment>;
}

/** The conservative default dial: reads auto, all writes gate. */
export const DEFAULT_DIAL: AutonomyDial = Object.freeze({
  writesMayAutoExecute: false,
  maxAutoBlastRadius: BlastRadius.Narrow,
  autoEnvironments: new Set([Environment.Dev, Environment.Staging]),
});

/**
 * The measured, deterministic inputs the policy function decides on. Everything
 * here is either graph/metadata-derived or a tighten-only signal — nothing the
 * model asserts can open the gate.
 */
export interface PolicyInput {
  effect: Effect; // catalog: read vs write
  actionCatalogued: boolean; // false => unknown action => fail safe
  blastRadius: BlastRadius; // graph
  reversibility: Reversibility; // catalog + target
  environment: Environment; // target metadata
  confidence: Confidence; // LLM, tighten-only
  ownerResolved: boolean; // graph
  ownerConfidence: Confidence; // graph edge confidence, tighten-only
  hasRollback: boolean; // write with no rollback => never auto
  /** Independent audit agent disagrees with the acting agent beyond threshold. */
  auditDivergence: boolean;
  /** Doc-grounded cautions. Tighten-only: presence can gate, absence never opens. */
  contextCautions: number;
}

const BLAST_ORDER: Record<BlastRadius, number> = {
  [BlastRadius.Narrow]: 0,
  [BlastRadius.Moderate]: 1,
  [BlastRadius.Wide]: 2,
};

function withinBlast(actual: BlastRadius, max: BlastRadius): boolean {
  return BLAST_ORDER[actual] <= BLAST_ORDER[max];
}

/**
 * The pure trust decision. Reason strings record every gate that fired so the
 * outcome is fully explainable and testable.
 *
 * Decision ordering is fail-safe: any single conservative trigger dominates.
 * The function only reaches AutoExecute after clearing *every* gate.
 */
export function policy(input: PolicyInput, dial: AutonomyDial = DEFAULT_DIAL): PolicyResult {
  const reasons: string[] = [];

  // Reads are autonomous. The real trust boundary is read vs. write.
  // A read must still be a known action; an unknown action is unknown risk.
  if (input.effect === Effect.Read) {
    if (!input.actionCatalogued) {
      return { decision: Decision.Escalate, reasons: ["unknown-action:fail-safe"] };
    }
    return { decision: Decision.AutoExecute, reasons: ["read:autonomous"] };
  }

  // ----- everything below is a WRITE -----

  // Fail-safe: an action the catalog has never classified has unknown risk.
  if (!input.actionCatalogued) {
    return { decision: Decision.Escalate, reasons: ["unknown-action:fail-safe"] };
  }

  // Independent audit disagreement is itself a deterministic escalation trigger.
  if (input.auditDivergence) {
    return { decision: Decision.Escalate, reasons: ["audit-divergence:escalate"] };
  }

  // No safe place to send it: unresolved or low-confidence ownership must never
  // be silently auto-routed. Missing data degrades to "ask a human".
  if (!input.ownerResolved) {
    return { decision: Decision.Escalate, reasons: ["owner-unresolved:escalate"] };
  }
  if (input.ownerConfidence === Confidence.Low) {
    return { decision: Decision.Escalate, reasons: ["owner-low-confidence:escalate"] };
  }

  // From here the fallback is a human owner's assent (RouteToOwner). We only
  // *downgrade* to AutoExecute if the change clears the full low-envelope bar.
  // Collect the reasons auto-execution is or isn't available.

  // Confidence is tighten-only: low confidence forces a human, but high
  // confidence is NEVER what authorizes auto-execution.
  const confidenceBlocksAuto = input.confidence === Confidence.Low;
  if (confidenceBlocksAuto) reasons.push("low-confidence:no-auto");

  // Context (untrusted docs) may only raise caution. Any doc-grounded caution
  // removes auto-eligibility; it can never lower the gate.
  const contextBlocksAuto = input.contextCautions > 0;
  if (contextBlocksAuto) reasons.push("context-caution:no-auto");

  // No rollback on a write => no autonomy, ever.
  if (!input.hasRollback) reasons.push("no-rollback:no-auto");

  // The envelope must be genuinely low: reversible + narrow-enough + non-prod.
  // These are the measured properties an attacker cannot fake.
  const reversibleEnough = input.reversibility === Reversibility.Reversible;
  if (!reversibleEnough) reasons.push("not-reversible:no-auto");

  const nonProd =
    input.environment !== Environment.Prod && input.environment !== Environment.Unknown;
  if (!nonProd) reasons.push("prod-or-unknown-env:no-auto");

  const blastOk = withinBlast(input.blastRadius, dial.maxAutoBlastRadius);
  if (!blastOk) reasons.push("blast-too-wide:no-auto");

  const envAllowed = dial.autoEnvironments.has(input.environment);
  if (!envAllowed) reasons.push("env-not-in-dial:no-auto");

  const dialOn = dial.writesMayAutoExecute;
  if (!dialOn) reasons.push("autonomy-dial-off:no-auto");

  const autoEligible =
    dialOn &&
    !confidenceBlocksAuto &&
    !contextBlocksAuto &&
    input.hasRollback &&
    reversibleEnough &&
    nonProd &&
    blastOk &&
    envAllowed;

  if (autoEligible) {
    return { decision: Decision.AutoExecute, reasons: ["low-envelope:earned-auto"] };
  }

  reasons.push("route-to-owner:human-assent");
  return { decision: Decision.RouteToOwner, reasons };
}
